java-checks/src/main/java/org/sonar/java/syntaxtoken/LastSyntaxTokenFinder.java

```diff
@@ -187,17 +231,215 @@ public void visitVariable(VariableTree tree) {
 
   @Override
   public void visitEnumConstant(EnumConstantTree tree) {
-    SyntaxToken identifierToken = tree.simpleName().identifierToken();
-    ClassTree enumConstantBody = ((NewClassTree) tree.initializer()).classBody();
-    if (enumConstantBody != null) {
-      scan(enumConstantBody);
-    } else {
-      lastSyntaxToken = identifierToken;
-    }
+    scan(tree.initializer());
   }
 
   @Override
   public void visitCaseLabel(CaseLabelTree tree) {
     lastSyntaxToken = tree.colonToken();
   }
+
+  @Override
+  public void visitBinaryExpression(BinaryExpressionTree tree) {
+    scan(tree.rightOperand());
+  }
+
+  @Override
+  public void visitLiteral(LiteralTree tree) {
+    lastSyntaxToken = tree.token();
+  }
+
+  @Override
+  public void visitIdentifier(IdentifierTree tree) {
+    lastSyntaxToken = tree.identifierToken();
+  }
+
+  @Override
+  public void visitCompilationUnit(CompilationUnitTree tree) {
+    if (!tree.types().isEmpty()) {
+      scan(Iterables.getLast(tree.types()));
+    } else if (!tree.imports().isEmpty()) {
+      scan(Iterables.getLast(tree.imports()));
+    } else {
+      // TODO(SONARJAVA-547) Should be the semi-colon token
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)